File: 0904-leaf-similar-trees/0904-leaf-similar-trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private List<Integer> leafs1 = new ArrayList();
    private List<Integer> leafs2 = new ArrayList();

    private void inOrder(TreeNode root, boolean isOne){
        if(root == null){
            return;
        }
        if(root.left == null && root.right == null){
            if(isOne){
                leafs1.add(root.val);  
            }else{
                leafs2.add(root.val);  
            }
        }else{
            inOrder(root.left, isOne);
            inOrder(root.right, isOne);
        }

    }
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        inOrder(root1, true);
        inOrder(root2, false);
        System.out.println(leafs1);
        System.out.println(leafs2);

        return leafs1.equals(leafs2);
    }
}

File: 0904-leaf-similar-trees/0904-leaf-similar-trees.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    val leafs1 =  mutableListOf<Int>()
    val leafs2 = mutableListOf<Int>()

    fun dfs(root: TreeNode?, leafs : MutableList<Int>){
        if(root == null){
            return
        }

        if(root?.left == null && root?.right == null){
            leafs.add(root.`val`)
        }else{
            dfs(root.left, leafs)
            dfs(root.right, leafs)
        }
    }
    fun leafSimilar(root1: TreeNode?, root2: TreeNode?): Boolean {
        dfs(root1, leafs1)
        dfs(root2, leafs2)

        return leafs1 == leafs2
    }
}